ec2/ec2test/server.go

```diff
@@ -2163,6 +2172,249 @@ func (srv *Server) unassignPrivateIP(w http.ResponseWriter, req *http.Request, r
 	}
 }
 
+type volume struct {
+	ec2.Volume
+}
+
+func (v *volume) matchAttr(attr, value string) (ok bool, err error) {
+	if strings.HasPrefix(attr, "attachment.") {
+		return false, fmt.Errorf("%q filter is not implemented", attr)
+	}
+	switch attr {
+	case "volume-type":
+		return v.VolumeType == value, nil
+	case "status":
+		return v.Status == value, nil
+	case "volume-id":
+		return v.Id == value, nil
+	case "size":
+		size, err := strconv.Atoi(value)
+		if err != nil {
+			return false, err
+		}
+		return v.Size == size, nil
+	case "availability-zone":
+		return v.AvailZone == value, nil
+	case "snapshot-id":
+		return v.SnapshotId == value, nil
+	case "encrypted":
+		encrypted, err := strconv.ParseBool(value)
+		if err != nil {
+			return false, err
+		}
+		return v.Encrypted == encrypted, nil
+	case "tag", "tag-key", "tag-value", "create-time":
+		return false, fmt.Errorf("%q filter is not implemented", attr)
+	}
+	return false, fmt.Errorf("unknown attribute %q", attr)
+}
+
+func (srv *Server) createVolume(w http.ResponseWriter, req *http.Request, reqId string) interface{} {
+	if req.Form.Get("AvailabilityZone") == "" {
+		fatalf(400, "MissingParameter", "missing availability zone")
+	}
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	ec2Vol := srv.parseVolume(req)
+	v := &volume{ec2Vol}
+	srv.volumes[v.Id] = v
+	var resp struct {
+		XMLName xml.Name
+		ec2.CreateVolumeResp
+	}
+	resp.XMLName = xml.Name{defaultXMLName, "CreateVolumeResponse"}
+	resp.RequestId = reqId
+	resp.Volume = v.Volume
+	return resp
+}
+
+func (srv *Server) parseVolume(req *http.Request) ec2.Volume {
+	volume := ec2.Volume{}
+	for attr, vals := range req.Form {
+		switch attr {
+		case "AWSAccessKeyId", "Action", "Signature", "SignatureMethod", "SignatureVersion",
+			"Version", "Timestamp":
+			continue
+		case "AvailabilityZone":
+			v := vals[0]
+			if v == "" {
+				fatalf(400, "MissingParameter", "missing availability zone")
+			}
+			volume.AvailZone = v
+		case "SnapshotId":
+			volume.SnapshotId = vals[0]
+		case "VolumeType":
+			volume.VolumeType = vals[0]
+		case "Size":
+			volume.Size = atoi(vals[0])
+		case "Iops":
+			volume.IOPS = int64(atoi(vals[0]))
+		case "Encrypted":
+			val, err := strconv.ParseBool(vals[0])
+			if err != nil {
+				fatalf(400, "InvalidParameterValue", "bad flag %s: %s", attr, vals[0])
+			}
+			volume.Encrypted = val
+		default:
+			fatalf(400, "InvalidParameterValue", "unknown field %s: %s", attr, vals[0])
+		}
+	}
+	volume.Id = fmt.Sprintf("vol-%d", srv.volumeId.next())
+	volume.Status = "available"
+	volume.CreateTime = time.Now().Format(time.RFC3339)
+	if volume.VolumeType == "" {
+		volume.VolumeType = "magnetic"
+	}
+	if volume.Size == 0 {
+		volume.Size = 1
+	}
+	return volume
+}
+
+func (srv *Server) deleteVolume(w http.ResponseWriter, req *http.Request, reqId string) interface{} {
+	v := srv.volume(req.Form.Get("VolumeId"))
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	delete(srv.volumes, v.Id)
```

Comment:
Should we also delete attachments here? What's the behaviour of EC2 if you try to delete a volume with attachments?